### PR TITLE
Fix issue #1265

### LINKED ISSUE #1265: Segfault when using generator in nopython mode
This code segfaults if I create an instance of the generator from Python before using it in a nopython mode function:
```python
from numba import jit
import numpy as np

@jit
def complex_grid(rmin, rmax, nr, imin, imax, ni):
    for r in np.linspace(rmin, rmax, nr):
        for i in np.linspace(imin, imax, ni):
            yield complex(r, i)

#### COMMENTING OUT THE NEXT TWO LINES FIXES THE SEGFAULT
for z in complex_grid(-2, 2, 100, -2, 2, 100):
    pass

@jit(nopython=True)
def mag2(z):
    return z.real * z.real + z.imag * z.imag

MAX_ITERS=1000

@jit(nopython=True)
def mandel(c):
    z = 0j
    for i in range(MAX_ITERS):
        z = z*z + c
        if mag2(z) >= 4:
            return 255 * i // MAX_ITERS
    return 255

@jit(nogil=False, nopython=True)
def mandel_patch(args):
    rmin, rmax, nr, imin, imax, ni = args
    points = np.empty(nr*ni, dtype=np.complex128)
    values = np.empty(nr*ni, dtype=np.uint8)

    for i, c in enumerate(complex_grid(rmin, rmax, nr, imin, imax, ni)):
        points[i] = c
        values[i] = mandel(c)
        
    return points, values

n = 1000
patches = [(-2, -1, n, -1, 0, n), (-1, 0, n, -1, 0, n), (0, 1, n, -1, 0, n),
           (-2, -1, n,  0, 1, n), (-1, 0, n,  0, 1, n), (0, 1, n,  0, 1, n)]

print(mandel_patch(patches[0]))
```

### PR DESCRIPTION
**Pull request by sklam**
* Fixed #1265
* Add tests for #1265
* Removed old ad-hoc handling of refcount in generator finalizer

### CODE REVIEW

File: numba/generators.py [hunk 1]
```diff
@@ -326,7 +316,14 @@ def lower_yield_suspend(self):
                                               0, state_index)
             ty = self.gentype.state_types[state_index]
             val = self.lower.loadvar(name)
+            # Load and DecRef previously stored value
+            oldval = self.context.unpack_value(self.builder, ty, state_slot)
+            if self.context.enable_nrt:
+                self.lower.decref(ty, oldval)
```
Comment (pitrou) [inline numba/generators.py]: If oldval and val are the same, decref before incref risks destroying the underlying data, no?

File: numba/tests/test_generators.py [hunk 2]
```diff
@@ -387,6 +387,34 @@ def test_nrt_gen0_no_iter(self):
...
+        py_res, py_old = factory(gen0)(arr)
+        c_gen = jit(nopython=True)(factory(jit(nopython=True)(gen0)))
+        c_res, c_old = c_gen(arr)
+
+        np.testing.assert_equal(py_res, c_res)
+
+        self.assertEqual(sys.getrefcount(py_res),
+                         sys.getrefcount(c_res))
+
+        self.assertEqual(sys.getrefcount(py_old),
```
Comment (pitrou) [inline numba/tests/test_generators.py]: py_old and c_old are the same object, no? Perhaps you want to call arr.copy()?

Comment (sklam) [inline numba/tests/test_generators.py]: Good catch.  Doing so revealed a refcount bug

**Comment by sklam** [comment]
There is a new expectedFailure due to https://github.com/numba/numba/issues/1278

**Pull request merged**